# **Task 1**
# Dataset Understanding

In [1]:
from google.colab import files
uploaded=files.upload()

Saving OnlineRetail.csv to OnlineRetail.csv


Importing several libraries pandas  for  data manuplation and analysis,numpy  for numerical values, matplotlib for     creatin visualization  and seaborn  for statistical analysis

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
df=pd.read_csv('/content/OnlineRetail.csv',encoding='latin1')
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850.0,United Kingdom


In [4]:

# Display first 5 rows
first_5_rows = df.head()
print("First 5 Rows:")
print(first_5_rows)

# Number of rows and columns
dataset_shape = df.shape
print("\nDataset Shape:")
print(dataset_shape)

# Data types of all columns
data_type = df.dtypes
print("\nData Types:")
print(data_type)

# Identify categorical features
categorical_features = df.select_dtypes(
    include=['object', 'category']
).columns

print("\nCategorical Features:")
print(categorical_features.tolist())

# Identify numerical features
numerical_features = df.select_dtypes(
    include=['int64', 'float64']
).columns

print("\nNumerical Features:")
print(numerical_features.tolist())

First 5 Rows:
  InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom  

Dataset Shape:
(541909, 8)

Data Types:
InvoiceNo       object
StockCode       object
Description     object
Quantity         int64
InvoiceDate     object
UnitPrice      float64
CustomerID     float



The Online Retail dataset contains **541,909 transaction records and 8 columns**. The dataset includes categorical features such as `InvoiceNo`, `StockCode`, `Description`, `InvoiceDate`, and `Country`, while `Quantity`, `UnitPrice`, and `CustomerID` are identified as numerical features. The `InvoiceDate` column is currently stored as an **object** data type and may need to be converted to datetime format during data cleaning.


# Dataset Overview

The dataset contains online retail transaction records from a UK-based e-commerce company.

Each transaction includes information such as:

- Invoice details
- Product information
- Quantity purchased
- Unit price
- Customer identification
- Country of purchase

These records provide valuable information regarding customer behavior, product demand, and revenue generation.

Before conducting advanced analysis, it is important to understand the dataset structure and quality.

# **TASK 2**
# Data Cleaning

In [5]:
#columns with missing values and number of missing values per column
df.isnull().sum()

,0
InvoiceNo,0
StockCode,0
Description,1454
Quantity,0
InvoiceDate,0
UnitPrice,0
CustomerID,135080
Country,0



Missing values in the `Description` column were replaced with **"Unknown"** because it is a categorical text field, and this method preserves all transaction records without creating inaccurate product descriptions. Missing values in the `CustomerID` column were replaced with **0** to represent unavailable customer information while retaining the transaction records for further analysis. This method was chosen to avoid deleting a large number of records, especially since `CustomerID` has **135,080 missing values**.


In [7]:
# Count duplicate rows before removal
duplicates_before = df.duplicated().sum()

# Remove duplicate rows
df = df.drop_duplicates()

# Display the number of rows removed
print("Number of duplicate rows removed:", duplicates_before)

# Display the remaining number of rows
print("Number of rows remaining:", len(df))

# Verify that no duplicates remain
print("Number of duplicate rows after removal:", df.duplicated().sum())

Number of duplicate rows removed: 5268
Number of rows remaining: 536641
Number of duplicate rows after removal: 0


In [8]:
# Standardize date format
df['InvoiceDate'] = pd.to_datetime(df['InvoiceDate'], errors='coerce')

# Standardize text formatting
df['Description'] = df['Description'].str.strip().str.upper()
df['Country'] = df['Country'].str.strip().str.title()

# Standardize column names
df.columns = df.columns.str.strip().str.lower().str.replace(' ', '_')

# Check data types
print(df.dtypes)

invoiceno              object
stockcode              object
description            object
quantity                int64
invoicedate    datetime64[ns]
unitprice             float64
customerid            float64
country                object
dtype: object


The InvoiceDate column was converted to a consistent datetime format, text formatting was standardized, column names were converted to lowercase with underscores, and data types were checked to ensure consistency.

In [9]:
# Check for invalid values
print("Negative Quantity:", (df['quantity'] < 0).sum())
print("Negative UnitPrice:", (df['unitprice'] < 0).sum())

# Check for inconsistent records
print("Zero Quantity:", (df['quantity'] == 0).sum())
print("Zero UnitPrice:", (df['unitprice'] == 0).sum())

# Check for incorrect date formats
print("Invalid InvoiceDate:", df['invoicedate'].isnull().sum())

Negative Quantity: 10587
Negative UnitPrice: 2
Zero Quantity: 0
Zero UnitPrice: 2510
Invalid InvoiceDate: 0


| Issue Found     | Action Taken                                                                                                            |
| --------------- | ----------------------------------------------------------------------------------------------------------------------- |
| Missing Values  | Filled missing `Description` values with **"Unknown"** and missing `CustomerID` values with **0**                       |
| Duplicates      | **5,268 duplicate rows removed**                                                                                        |
| Invalid Entries | Negative quantities and prices, as well as zero-price transactions, were identified and reviewed as potential anomalies |
| Standardization | Date formats, text formatting, column names, and data types were standardized                                           |


# **TASK 3**

# Exploratory Data Analysis (EDA)

In [10]:
df.describe()

,quantity,invoicedate,unitprice,customerid
count,536641.000000,536641,536641.000000,536641.000000
mean,9.620029,2011-07-04 08:57:06.087421952,4.632656,11435.904653
min,-80995.000000,2010-12-01 08:26:00,-11062.060000,0.000000
25%,1.000000,2011-03-28 10:52:00,1.250000,0.000000
50%,3.000000,2011-07-19 14:04:00,2.080000,14336.000000
75%,10.000000,2011-10-18 17:05:00,4.130000,16241.000000
max,80995.000000,2011-12-09 12:50:00,38970.000000,18287.000000
std,219.130156,NaN,97.233118,6795.044250


# **Top 10 Countries by Number of Transactions**

In [16]:
country_transactions = (
    df['country']
    .value_counts()
    .head(10)
    .reset_index()
    )
print(country_transactions)

          country   count
0  United Kingdom  490300
1         Germany    9480
2          France    8541
3            Eire    8184
4           Spain    2528
5     Netherlands    2371
6         Belgium    2069
7     Switzerland    1994
8        Portugal    1510
9       Australia    1258


In [17]:


country_transactions.columns = ['Country', 'Number of Transactions']

fig = px.bar(
    country_transactions,
    x='Country',
    y='Number of Transactions',
    color='Country',
    text='Number of Transactions',
    title='Top 10 Countries by Number of Transactions'
)

fig.update_traces(textposition='outside')

fig.update_layout(
    title_x=0.5,
    template='plotly_white',
    xaxis_title='Country',
    yaxis_title='Number of Transactions',
    showlegend=False
)

fig.show()

The analysis identifies the countries with the highest number of transactions. The United Kingdom has the largest number of transactions, indicating that it is the primary market for the online retail business. Other countries in the top 10 represent additional international markets where the business has a significant customer presence.

# **Monthly Sales Trend**

In [19]:
df['Sales'] = df['quantity'] * df['unitprice']

In [30]:
monthly_sales =(
     df.groupby(df['invoicedate'].dt.to_period('M'))['Sales']
     .sum()
     .reset_index()
     )
print(monthly_sales)

   invoicedate        Sales
0      2010-12   746723.610
1      2011-01   558448.560
2      2011-02   497026.410
3      2011-03   682013.980
4      2011-04   492367.841
5      2011-05   722094.100
6      2011-06   689977.230
7      2011-07   680156.991
8      2011-08   681386.460
9      2011-09  1017596.682
10     2011-10  1069368.230
11     2011-11  1456145.800
12     2011-12   432701.060


In [33]:
monthly_sales['invoicedate'] = monthly_sales['invoicedate'].astype(str)


fig = px.line(
    monthly_sales,
    x='invoicedate',
    y='Sales',
    title='Monthly Sales Trend',
    markers=True
)

fig.update_layout(
    title_x=0.5,
    template='plotly_white',
    xaxis_title='Month',
    yaxis_title='Total Sales'
)

fig.show()



The monthly sales trend shows that sales **fluctuated throughout the year**. Sales reached their **highest point in November 2011**, at approximately **1.45 million**, indicating a significant increase in customer purchasing activity. The lowest sales were recorded in **December 2011**, at approximately **0.43 million**, showing a sharp decline after the November peak. Overall, sales increased significantly toward the end of the year, particularly from **August to November 2011**.


# **Top 10 Best-Selling Products**

In [36]:
top_products = (
    df.groupby('description')['quantity']
    .sum()
    .sort_values(ascending=False)
    .head(10)
    .reset_index())
print(top_products)

                          description  quantity
0   WORLD WAR 2 GLIDERS ASSTD DESIGNS     53751
1             JUMBO BAG RED RETROSPOT     47260
2                      POPCORN HOLDER     36322
3       ASSORTED COLOUR BIRD ORNAMENT     36282
4     PACK OF 72 RETROSPOT CAKE CASES     36016
5  WHITE HANGING HEART T-LIGHT HOLDER     35298
6                  RABBIT NIGHT LIGHT     30631
7              MINI PAINT SET VINTAGE     26437
8           PACK OF 12 LONDON TISSUES     26299
9  PACK OF 60 PINK PAISLEY CAKE CASES     24719


In [37]:

top_products.columns = ['Product', 'Quantity Sold']

fig = px.bar(
    top_products,
    x='Quantity Sold',
    y='Product',
    color='Product',
    text='Quantity Sold',
    title='Top 10 Best-Selling Products',
    orientation='h'
)

fig.update_traces(textposition='outside')

fig.update_layout(
    title_x=0.5,
    template='plotly_white',
    xaxis_title='Quantity Sold',
    yaxis_title='Product',
    showlegend=False
)

fig.show()



The analysis shows that **WORLD WAR 2 GLIDERS ASSTD DESIGNS** is the best-selling product, with **53,751 units sold**, followed by **JUMBO BAG RED RETROSPOT** with **47,260 units**. **PACK OF 60 PINK PAISLEY CAKE CASES** recorded the lowest quantity among the top 10 products, with **24,719 units sold**. Overall, the results show that these products have significantly higher demand and contribute strongly to the total sales volume.


# **Number of Customers by Country**

In [45]:
customers_by_country = (
    df.groupby('country')['customerid']
    .nunique()
    .sort_values(ascending=False)
    .head(10)
    .reset_index())
print(customers_by_country)

          country  customerid
0  United Kingdom        3951
1         Germany          95
2          France          88
3           Spain          31
4         Belgium          25
5     Switzerland          22
6        Portugal          20
7           Italy          15
8         Finland          12
9         Austria          11


In [47]:
customers_by_country.columns = ['Country', 'Number of Customers']


fig = px.bar(
    customers_by_country,
    x='Country',
    y='Number of Customers',
    color='Country',
    text='Number of Customers',
    title='Top 10 Countries by Number of Customers'
)

fig.update_traces(textposition='outside')

fig.update_layout(
    title_x=0.5,
    template='plotly_white',
    xaxis_title='Country',
    yaxis_title='Number of Customers',
    showlegend=False
)

fig.show()



The **United Kingdom** has the highest number of customers, with **3,951 customers**, significantly exceeding all other countries. **Germany** follows with **95 customers**, while **France** has **88 customers**. The remaining countries have fewer than 31 customers each. This indicates that the business has a **strong customer concentration in the United Kingdom**, with a much smaller customer base across other international markets.


# **Quantity vs Unit Price**

In [50]:
fig = px.scatter(
    df,
    x='unitprice',
    y='quantity',
    title='Relationship Between Quantity and Unit Price',
    opacity=0.5
)

fig.update_layout(
    title_x=0.5,
    template='plotly_white',
    xaxis_title='Unit Price',
    yaxis_title='Quantity Purchased'
)

fig.show()

The scatter plot shows that most transactions are concentrated around low unit prices and smaller quantities purchased. However, there are several extreme values and outliers, including unusually high quantities and unit prices. The presence of negative quantities and prices also suggests possible returns, cancellations, or data anomalies that may require further investigation.

## Key Insights

* The **United Kingdom** is the primary market, accounting for the largest share of transactions in the dataset.
* The business has a strong **international customer presence**, with transactions recorded across multiple countries.
* The **monthly sales trend** shows changes in business activity over time, highlighting periods of higher and lower sales performance.
* A small number of products account for a large share of the **total quantity sold**, indicating that certain products are significantly more popular than others.
* The **Quantity vs Unit Price** analysis reveals several extreme values and outliers, including negative quantities and unusually high prices, which may be associated with returns, cancellations, or data anomalies.
* The dataset contains **5,268 duplicate records**, which were removed during data cleaning to improve the reliability of the analysis.
* Overall, the analysis provides insights into **sales trends, customer markets, product demand, and unusual transactions**, supporting better understanding of the retail business.
